src/infra/flow-system/inmem/src/flow/flow_event_store_inmem.rs

```diff
@@ -343,6 +343,34 @@ impl FlowEventStore for FlowEventStoreInMem {
         Box::pin(futures::stream::iter(flow_ids_page))
     }
 
+    #[tracing::instrument(level = "debug", skip_all)]
+    fn get_all_flow_ids_by_datasets(
+        &self,
+        dataset_ids: Vec<DatasetID>,
+        pagination: FlowPaginationOpts,
+    ) -> (usize, FlowIDStream) {
+        let mut total_count = 0;
+        let flow_ids_page: Vec<_> = {
+            let state = self.inner.as_state();
+            let g = state.lock().unwrap();
+            let mut result: Vec<Result<FlowID, _>> = vec![];
+            g.all_flows.iter().rev().for_each(|flow_id| {
+                let flow_key = g.flow_key_by_flow_id.get(flow_id).unwrap();
+                if let FlowKey::Dataset(flow_key_dataset) = flow_key {
+                    if dataset_ids.contains(&flow_key_dataset.dataset_id) {
+                        if total_count >= pagination.offset && result.len() <= pagination.limit {
+                            result.push(Ok(*flow_id));
+                        }
+                        total_count += 1;
```

Comment:
If "count" and "list" are separated, you can break the loop as soon as "limit" is exhausted.